src/Sarif/Baseline/V2/V2ResultMatcher.cs

```diff
@@ -194,6 +231,54 @@ private void LinkIfSimilar(int beforeIndex, int afterIndex)
             }
         }
 
+        private ExtractedResult FirstWithUri(ExtractedResult desiredUri, IList<ExtractedResult> set, ref int fromIndex)
+        {
+            // Find the first Result with a Uri *matching* the desired one, or null if there aren't any
+            for (; fromIndex < set.Count; ++fromIndex)
```

Comment:
>fromIndex [](start = 44, length = 9)

Do we have some N-squared action going on here? The caller is looping over the list same list. If so, can we avoid it by doing a pre-scan of the two lists, finding the first Uri in each element of each list, and making some dictionaries? Something like `Dictionary<Uri, int> firstBeforeUriIndexDictionary, lastBeforeUriIndexDictionary, firstAfterUriIndexDictionary, lastAfterUriIndexDictionary`?

Reply:
No, but it's tricky. Added some comments to explain.

Basically, LinkFirstAndLastFromSameArtifact is similar to LinkResultsWithIdenticalWhere in that it walks the two lists together in a shared loop and always moves one or the other forward.

However, it's too complex for one method, so I keep two indices (beforeIndex and afterIndex) and pass them by ref to FirstWithUri and LastWithUri. Neither of them reset the indices, so they are always walking forward in the two sets.